### PR TITLE
Obstructing view if dismiss popover region apple provides in UIPopove…

### PR DESCRIPTION
**Pull request by rafael-chavez**
…rController is not in the tree.

### CODE REVIEW

**Comment by rafael-chavez** [comment]
@marekcirkos

File: WebDriverAgentLib/FBAlert.m [hunk 1]
```diff
@@ -44,8 +44,16 @@ - (XCUIElement *)fb_alertElement
...
+  //Only return sheet if it does not have a dismiss popover region.
+  BOOL isIpad = [UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad;
+  //check for Apple's popover region in a UIPopoverController.
+  NSPredicate *predicateString = [NSPredicate predicateWithFormat:@"identifier == 'PopoverDismissRegion'"];
+  XCUIElementQuery *query = [[self.application descendantsMatchingType:XCUIElementTypeAny] matchingPredicate:predicateString];
+  NSArray *childElements = [query allElementsBoundByIndex];
+    
   alert = self.sheets.element;
-  if (alert.exists) {
+  if (alert.exists && !isIpad && childElements.count == 0) {
```
Comment (marekcirkos) [inline WebDriverAgentLib/FBAlert.m]: This will not return any sheet on iPad

**Comment by marekcirkos** [comment]
Please also add integration test to `FBAlertTests`

File: WebDriverAgentLib/FBAlert.m [hunk 2]
```diff
@@ -44,8 +44,16 @@ - (XCUIElement *)fb_alertElement
...
+  //Only return sheet if it does not have a dismiss popover region.
+  BOOL isIpad = [UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad;
+  //check for Apple's popover region in a UIPopoverController.
+  NSPredicate *predicateString = [NSPredicate predicateWithFormat:@"identifier == 'PopoverDismissRegion'"];
+  XCUIElementQuery *query = [[self.application descendantsMatchingType:XCUIElementTypeAny] matchingPredicate:predicateString];
+  NSArray *childElements = [query allElementsBoundByIndex];
+    
   alert = self.sheets.element;
-  if (alert.exists) {
+  if (alert.exists && isIpad && childElements.count == 0) {
```
Comment (rafael-chavez) [inline WebDriverAgentLib/FBAlert.m]: it should be isPad not !isIpad.Thanks @marekcirkos

**Comment by facebook-github-bot-2** [comment]
@rafael-chavez updated the pull request.

**Comment by facebook-github-bot-2** [comment]
@rafael-chavez updated the pull request.

File: WebDriverAgentLib/FBAlert.m [hunk 3]
```diff
@@ -44,8 +44,16 @@ - (XCUIElement *)fb_alertElement
...
+  //Only return sheet if it does not have a dismiss popover region.
+  BOOL isIpad = [UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad;
+  //check for Apple's popover region in a UIPopoverController.
+  NSPredicate *predicateString = [NSPredicate predicateWithFormat:@"identifier == 'PopoverDismissRegion'"];
+  XCUIElementQuery *query = [[self.application descendantsMatchingType:XCUIElementTypeAny] matchingPredicate:predicateString];
+  NSArray *childElements = [query allElementsBoundByIndex];
+    
   alert = self.sheets.element;
-  if (alert.exists) {
+  if ((alert.exists && isIpad && childElements.count == 0) || (alert.exists && !isIpad)) {
```
Comment (rafael-chavez) [inline WebDriverAgentLib/FBAlert.m]: it should be isPad not !isIpad. Popover only exists on ipad.thanks @marekcirkos !

Comment (marekcirkos) [inline WebDriverAgentLib/FBAlert.m]: This is hard to read what about `alert.exists && (!isIPad || childElements.count == 0)`?

Comment (rafael-chavez) [inline WebDriverAgentLib/FBAlert.m]: I could also do it like:

```
  if (alert.exists && ((isIpad && childElements.count == 0) || !isIpad)) 
```

not sure if you have a pref.

Comment (rafael-chavez) [inline WebDriverAgentLib/FBAlert.m]: hahha, i think we put that at the same time.

Comment (rafael-chavez) [inline WebDriverAgentLib/FBAlert.m]: updating...

**Comment by facebook-github-bot-2** [comment]
@rafael-chavez updated the pull request.

**Comment by rafael-chavez** [comment]
Adding integration tests...

File: WebDriverAgentLib/FBAlert.m [hunk 4]
```diff
@@ -44,8 +44,16 @@ - (XCUIElement *)fb_alertElement
...
+  //Only return sheet if it does not have a dismiss popover region.
+  BOOL isIpad = [UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad;
+  //check for Apple's popover region in a UIPopoverController.
+  NSPredicate *predicateString = [NSPredicate predicateWithFormat:@"identifier == 'PopoverDismissRegion'"];
+  XCUIElementQuery *query = [[self.application descendantsMatchingType:XCUIElementTypeAny] matchingPredicate:predicateString];
+  NSArray *childElements = [query allElementsBoundByIndex];
+    
   alert = self.sheets.element;
-  if (alert.exists) {
+  if (alert.exists && ((isIpad && childElements.count == 0) || !isIpad)) {
```
Comment (marekcirkos) [inline WebDriverAgentLib/FBAlert.m]: I think the second && is not needed.

**Comment by marekcirkos** [comment]
I just landed running tests on iPhone and iPad with Travis. You can rebase to check your changes ;)

**Comment by rafael-chavez** [comment]
updated @marekcirkos added integration test for ipad.

**Comment by facebook-github-bot-2** [comment]
@rafael-chavez updated the pull request.

**Comment by facebook-github-bot-2** [comment]
@rafael-chavez updated the pull request.

File: WebDriverAgentTests/IntegrationTests/FBAlertTests.m [hunk 5]
```diff
@@ -124,4 +130,18 @@ - (void)testGPSAccessAlert
   XCTAssertTrue([alert.text containsString:@"Yo Yo"]);
 }
 
+- (void)testSheetAlert
+{
+    FBAlert *alert = [FBAlert alertWithApplication:self.testedApplication];
+    BOOL isIpad = [UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad;
+    //On iphone this filterObstructedElements will throw an exception.
```
Comment (marekcirkos) [inline WebDriverAgentTests/IntegrationTests/FBAlertTests.m]: Could you add another test testing if iPhone will throw?

Comment (rafael-chavez) [inline WebDriverAgentTests/IntegrationTests/FBAlertTests.m]: will do.

**Comment by rafael-chavez** [comment]
added iphone as well @marekcirkos

**Comment by facebook-github-bot-2** [comment]
@rafael-chavez updated the pull request.

**Comment by facebook-github-bot-2** [comment]
@rafael-chavez updated the pull request.

File: WebDriverAgentTests/IntegrationTests/FBAlertTests.m [hunk 6]
```diff
@@ -124,4 +130,27 @@ - (void)testGPSAccessAlert
...
+  BOOL isIpad = [UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad;
+  [self showApplicationSheet];
+  XCUIElement *showSheetButton = self.testedApplication.buttons[FBShowSheetAlertButtonName];
+  //On iphone this filterObstructedElements will throw an exception.
+  if (isIpad) {
+    NSArray *filteredElements = [alert filterObstructedElements:@[showSheetButton]];
+    XCTAssertEqualObjects(filteredElements, @[showSheetButton]);
+  } else {
+    NSException *exception;
+    @try {
```
Comment (marekcirkos) [inline WebDriverAgentTests/IntegrationTests/FBAlertTests.m]: You might want to use `XCTAssertThrowsSpecificNamed` instead.

File: WebDriverAgentTests/IntegrationTests/FBAlertTests.m [hunk 7]
```diff
@@ -124,4 +130,19 @@ - (void)testGPSAccessAlert
...
+  FBAlert *alert = [FBAlert alertWithApplication:self.testedApplication];
+  BOOL isIpad = [UIDevice currentDevice].userInterfaceIdiom == UIUserInterfaceIdiomPad;
+  [self showApplicationSheet];
+  XCUIElement *showSheetButton = self.testedApplication.buttons[FBShowSheetAlertButtonName];
+  //On iphone this filterObstructedElements will throw an exception.
+  if (isIpad) {
+    NSArray *filteredElements = [alert filterObstructedElements:@[showSheetButton]];
+    XCTAssertEqualObjects(filteredElements, @[showSheetButton]);
+  } else {
+    XCTAssertThrowsSpecificNamed([alert filterObstructedElements:@[showSheetButton]], NSException, FBAlertObstructingElementException, @"should throw FBAlertObstructingElementException");
```
Comment (rafael-chavez) [inline WebDriverAgentTests/IntegrationTests/FBAlertTests.m]: Updated @marekcirkos

**Comment by facebook-github-bot-2** [comment]
@rafael-chavez updated the pull request.

**Comment by marekcirkos** [comment]
This is nice. Thank you for your PR and changes!

**Comment by marekcirkos** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-8** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/660577674099469/int_phab) to review.

**Pull request closed without merging**